### DISCUSSION TITLE
even when loading livewire scripts in head I get> "Uncaught TypeError: Cannot read property '$wire' of undefined"

### DISCUSSION THREAD

**Discussion by torian257x** [livewire/livewire#4357] [1. Help]
> Uncaught TypeError: Cannot read property '$wire' of undefined

I load livewire scripts in head

I load alpine js scripts at the BOTTOM of body like so:
```
<script defer src="{{ asset('plugins/alpine/alpine_collapse.js') }}"></script>
<script defer src="{{ asset('plugins/alpine/alpine.js') }}"></script>
```

my code:
```
<script type="text/javascript">
        document.addEventListener('alpine:init', () =>
        {
            Alpine.store('detailsStore', {
                alpineModalOpen: @entangle('showDetailViewModal'),
            });
        });
    </script>
```

as recommended in the docs https://alpinejs.dev/globals/alpine-store

is there something I can listen to, to properly load javascript? I hope there is some "secret" because this is a bit annoying.

**Comment by joshhanley** [comment]
@torian257x hmm I don't think using entangle with an Alpine store property will work, which is probably why the `$wire` error is being thrown.

Not sure what you mean by "something I can listen to, to properly load javascript", can you elaborate?

**Reply by torian257x** [reply]
I thought its a loading problem e.g. of sequence e.g. somehow something is not loaded when it should be.

Why should it not work though?

the thing with entangle not working is a pity, because I use it heavily for my modals. As I want to open and close modals with ease... 
we use stores fairly heavily, I dont see why stores shouldnt be able to entangle?

Would be great to make that work. I thought the store is a first class citizen somehow...

I am now using events e.g. in component

```
        $this->dispatchBrowserEvent('openMyModal');
```


in blade

```
        document.addEventListener('openMyModal', () => {
            Alpine.store('detailsStore').alpineModalOpen = true;
        });
```

same for close.

**Comment by joshhanley** [comment]
@torian257x yeah I've tried to get it running before, but couldn't get it running.

Half of the issue is typically the store isn't necessarily set up within a Livewire component, and also how to handle multiple Livewire components entangling to the same store value?

I built Sprucewire to allow binding between Spruce (a store before Alpine had a first party one) and LW and ran into similar issues of how to bind correctly.

**Reply by torian257x** [reply]
@joshhanley entangling multiple times to the same thing should just be error or if not possible otherwise, malfunction. We don't support bad life choices, we really dont. 

for me the store is setup in the livewire component, but I see the problem. But it should be possible to access the right component anyway, via javascript? e.g. instead of @entangle that you can only use inside a component, outside a component they could use:

`Livewire.find([component-id]).entangle('count')` or something according to that. Maybe with actually easy to use component-id, not this magical hash as id nobody knows how to reproduce. E.g. I want to be able to set some id or key or whatever, that I then can easily access via the javascript method (maybe already possible, I just cant find the docs any more)

those should be fairly easy to solve problems imho. I'm not saying its not work, but the decisions look fairly straight forward to me.

**Reply by torian257x** [reply]
@joshhanley what you really need in livewire is a livewire global state in the backend. Not frontend (we have that already). But backend still limping behind. emit event emit event emit event then get the information from where? `session()` ? there is a puzzle piece missing in livewire, easy to spot if you look at vue store.

**Reply by joshhanley** [reply]
@torian257x hmm but if you don't want multiple things to be able to entangle, then I'd argue that you shouldn't be using the store for that instead just use an Alpine property. What do you think?

Yeah finding a component by `wire:key` would be interesting but they don't need to be unique so I don't think that'd work. By ID could be a good idea though, as for valid html they should be unique.

I actually think a global frontend store like Vuex but for Livewire could solve a lot of issues of sharing data between components, which saves firing events everywhere. Is there any reason in particular you think that should be on the backend instead?

**Reply by torian257x** [reply]
alpine property> alpine properties have heavy and strict limitations for reasons that aren't clear to me. First and foremost you cannot access and modify alpine properties from outside the component which is a clear anti pattern imho. And with anti pattern I mean: its an anti pattern you cannot access and write to the properties. This is very very cumbersome, and Im not trying to knock something or someone. 

But that is why I was so stoked when alpine 3 released their own store. Btw I am stoked as well about vue 3 state management and say, react zustand library. Those are all going in the right direction. Alpine 2 wasnt, neither was vuex or redux, thus the new popular stores.

I started writing my components using stores always, because for me ease of refactoring is important. E.g. if it costs you effort to refactor something, then you won't do it. Then you will end with trash code, because nobody ever refactored stuff. Refactoring with stores is easier, you can split apart, nest etc without visibility problems, call/read from outside, everything is better... in theory 

entangle multiple components can entangle> fair enough, I somewhat revised my opinion as well... I noticed that my use case is a bit limited. There speaks not much against entangling multiple components, apart from technicalities I guess.

Backend store> Im not sure I understand the question... I think livewire, generally, should give the option to have the state in the backend (but its not that high priority atm). Java's JSF has implementations of backend state (they do have client side state as well), even optimized separate ones. Not sure if it's helpful as PHP is tear down and build up whereas java isn't but might be worth a look, I think they serialize and unserialize anyway... When I was working for a big company they ALWAYS used backend state as they argued its more secure. And they are probably right with that. There is a mojarra implementation of the state management here afaik https://github.com/eclipse-ee4j/mojarra/blob/d3aa2e7ae5f7eaab52fbc077e234eb7189b8f356/impl/src/main/java/com/sun/faces/renderkit/ResponseStateManagerImpl.java 

What is more important is as you say, a livewire global state, if you transfer that state back and forth between client and server is less relevant I guess... But it's important that you can access that global state and the components get notified of changes to the state who consume it. 
Optimally keep the state on the server this way you dont need send back and forth data but I doubt that's a low hanging fruit, so... not so important.

having the state in the client on the other hand could be an advantage, if you can manipulate it or say, define pieces you can manipulate client side, would reduce A LOT of round trips potentially. But it should be "sign in" style e.g. point out state that can be changed client side. The current way it is done is imho fairly insecure e.g. you can change "some" (aka primitives, arrays... standard collections?) component state via client side requests but not all (models, eloquent collections?)
but lets not start a construction everywhere, just a "simple" livewire state would be great
